### PR TITLE
Add 2FA (OTP) #1049

### PR DESCRIPTION
**Pull request by brussens**
##### CHANGES PROPOSED:

-  Adds Two Factor Authentication (OTP)

###### ISSUES ADDRESSED:

https://github.com/azerothcore/azerothcore-wotlk/issues/1049

##### TESTS PERFORMED:

Tested on local windows server (WIN10, x64 PRO) and Production linux server Debian x64.

##### HOW TO TEST THE CHANGES:

Compile core, add !!!Base32!!! (16 digits) code to token_key field in account table, register this code in Google Authenticator and try login.

##### KNOWN ISSUES AND TODO LIST:

- [ ]
- [ ] 

##### Target branch(es):

Master

### CODE REVIEW

File: src/server/authserver/Server/TOTP.h [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+ * This program is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for
+ * more details.
+ *
+ * You should have received a copy of the GNU General Public License along
+ * with this program. If not, see <http://www.gnu.org/licenses/>.
+ */
+#ifndef _TOTP_H
+#define _TOPT_H
```
Comment (BarbzYHOOL) [inline src/server/authserver/Server/TOTP.h]: topt? typo

Comment (brussens) [reply src/server/authserver/Server/TOTP.h]: Hmm, yes) Carried everything in manual)

File: src/server/authserver/Server/TOTP.cpp [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+     if (count < bufSize)
+        result[count] = '\000';
+    return count;
+}
+#define HMAC_RES_SIZE 20
+namespace TOTP
+{
+    unsigned int GenerateToken(const char* b32key)
+    {
+        size_t keySize = strlen(b32key);
```
Comment (BarbzYHOOL) [inline src/server/authserver/Server/TOTP.cpp]: TOTP.cpp:64:26: fatal error: 
      use of undeclared identifier 'strlen'
        size_t keySize = strlen(b32key);

Comment (Sdyess) [reply src/server/authserver/Server/TOTP.cpp]: Just needs to add` #include <cstring>`

File: src/server/authserver/Server/AuthSocket.h [hunk 3]
```diff
@@ -67,6 +67,7 @@ class AuthSocket: public RealmSocket::Session
     eStatus _status;
 
     std::string _login;
+	std::string _tokenKey;
```
Comment (Sdyess) [inline src/server/authserver/Server/AuthSocket.h]: spacing is wrong

**Comment by Sdyess** [comment]
If someone is able to test this, this could probably be merged. I’ll see about setting up 2FA locally and checking it if no one else does it in the near future.

**Comment by BarbzYHOOL** [comment]
No idea how to test it but apparently brussens uses it on his own website, so at least 1 test has been done (by the PR's author).

By the way, isn't "TOTP" equal to "TrinityOTP" ? :D

**Comment by brussens** [comment]
I tested this on production and fresh local compiled server

**Comment by BarbzYHOOL** [comment]
@Sdyess no test? :'(

**Comment by Sdyess** [comment]
Thanksgiving break next week, remind me in a week 😎

**Comment by BarbzYHOOL** [comment]
@Sdyess now! :p

**Comment by BarbzYHOOL** [comment]
@Sdyess Have you tested a bit?

**Comment by BarbzYHOOL** [comment]
@brussens Thank you mate!!!

**Pull request merged**